torch/utils/_sympy/value_ranges.py

```diff
@@ -443,6 +443,22 @@ def floor(cls, x):
     def ceil(cls, x):
         return ValueRanges.increasing_map(x, sympy.functions.elementary.integers.ceiling)
 
+    @classmethod
+    def round(cls, number, ndigits=None):
+        if ndigits is not None:
+            assert ndigits.is_singleton()
+            ndigits = ndigits.lower
+
+        # Negative ndigits not just affect the decimal, but the integer part as well
+        if ndigits is None or ndigits >= 0:
+            lower = sympy.floor(number.lower)
```

Comment:
If `ndigits > = 0` this formula is not correct. I don't think this is a lower bound.

I think you need a formula similar to that of the negative case. 

Note that this is a increasing function, so if you implement it so that it works with sympy, you can simply call `ValueRanges.increasing_map`.

Reply:
> If `ndigits > = 0` this formula is not correct. I don't think this is a lower bound.
> 
> I think you need a formula similar to that of the negative case.

Why not? `math.floor(x) <= round(x, ndigits) <= math.ceil(x)` holds for `ndigits >= 0`. Non-negative `ndigits` just effect the decimal part.

> Note that this is a increasing function, so if you implement it so that it works with sympy, you can simply call `ValueRanges.increasing_map`.

Nope. That would require to apply the same function to both the lower and upper limit:

https://github.com/pytorch/pytorch/blob/55ce4693ff2c0b6e50b8af323f36ecc7ff929638/torch/utils/_sympy/value_ranges.py#L127-L131

But we need to floor the lower and ceil the upper.

Reply:
I think `increasing_map` is correct and gives a tighter bound but it's only a difference of 1, so not sure how big a deal it is. If you consider a lower bound of 0.8 which rounds up to 1, there is no value between 0.8 and infinity which rounds to anything less than 1. Similarly if you have an upper bound of 1.2 which rounds down to 1, there is no value between 1.2 and negative infinity that rounds to anything greater than 1. Therefore, the bound [1, 1] is correct but your formula gives [0, 2]. This is still valid though, it's just looser than necessary.

Reply:
@peterbell10 is right here, but I would argue that here being off by 1 is important, because these functions are used to prove that an iteration variable is <= the size they are iterating over. These iterations are often tight, so an off-by-one error is important.

Also, using the `increasing_map` here will avoid bugs like the one in the `else`  branch, where you use `floor` for the upper bound.

Reply:
I was not aware that we have access to an actual number here rather than just an expression. For the latter, I think we can't do better than what I have proposed. However, for the former, you are right that we can simply invoke `round` and thus have a tighter bound.